### PR TITLE
[B2BP-179] - Implementing Header as generic component

### PR DESCRIPTION
**Pull request by Meolocious**
#### List of Changes
- Created `Header.tsx` component inside the `components` folder to isolate the 'use client' side only for the component itself.
- Added `Header.ts` file for fetching data from the CMS API call, utilizing io-ts for validation.
- Introduced `Menu.ts` file to populate the menu part of the header with routes obtained from `navigation.ts`.
- Included `Header.test.ts` file for testing the API call using `vitest`.

#### Motivation and Context
The motivation behind these changes is to enhance the structure and maintainability of our codebase. By creating the `Header.tsx` component, we isolate client-side logic for better organization and modularity. The addition of `Header.ts` ensures a reliable and validated approach to fetching data from the CMS API, employing io-ts for robust data validation. The `menu.ts` file contributes to filling the header menu with routes obtained from `navigation.ts`, providing a seamless and dynamic user experience. The testing file, `Header.test.ts`, validates the API call using `vitest`, ensuring the functionality is robust and dependable.

#### How Has This Been Tested?
Comprehensive testing has been conducted to validate the effectiveness and reliability of these changes. The `Header.test.ts` file includes a suite of tests for the API call, covering various scenarios and edge cases. Additionally, the testing environment has been carefully configured to assess how these changes impact other areas of the code. This includes integration tests to ensure the seamless interaction between components, API calls, and data validation.

#### Types of changes
- [ ] Chore (nothing changes from a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Doesn't need any documentation updates.

### CODE REVIEW

File: apps/nextjs-website/src/lib/fetch/types/ctaButtons.ts
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/types/ctaButtons.ts:1]: Rename the file to `CtaButton.ts`

File: apps/nextjs-website/src/lib/fetch/types/ctaButtons.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  }),
+  t.partial({
+    icon: t.union([t.string, t.null]),
+    size: CTAButtonSize,
+  }),
+]);
+
+export type ctaButtons = t.TypeOf<typeof CtaButtonsSchema>;
+
+export const CtaGroupCodec = t.intersection([
```
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/types/ctaButtons.ts:41]: Move CtaGroupCodec to `src/lib/fetch/types/CtaGroup.ts`

File: apps/nextjs-website/src/lib/fetch/menu.ts
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/menu.ts:1]: Move this function to `src/lib/menu.ts` because it does not use `fetch`

Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/menu.ts:1]: Add some test to `makeMenuFromNavigation`

**Review by datalek — CHANGES_REQUESTED** [review]

File: apps/nextjs-website/src/lib/fetch/__tests__/header.test.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          href: '/',
+          variant: 'text',
+          color: 'inherit',
+          icon: null,
+          size: 'medium',
+        },
+      ],
+    },
+  },
+};
```
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/__tests__/header.test.ts:45]:
```suggestion
const headerResponse = {
  data: {
    attributes: {
      productName: 'Test',
      beta: false,
      reverse: false,
      theme: 'light',
      avatar: {
        data: {
          attributes: {
            url: '/uploads/file_e4c03befba.png',
            alternativeText: null,
          },
        },
      },
      ctaButtons: [
        {
          text: 'test',
          href: '/',
          variant: 'text',
          color: 'inherit',
          icon: null,
          size: 'medium',
        },
      ],
    },
  },
};
```

File: apps/nextjs-website/src/lib/fetch/types/StrapiImage.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  data: t.union([
+    t.type({
+      attributes: t.type({
+        alternativeText: t.union([t.string, t.null]),
+        url: t.string,
+      }),
+    }),
+    t.null,
+  ]),
+});
```
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/types/StrapiImage.ts:13]: Use strict to have the output that contains only the defined attributes

```suggestion
export const StrapiImageSchema = t.strict({
  data: t.union([
    t.strict({
      attributes: t.strict({
        alternativeText: t.union([t.string, t.null]),
        url: t.string,
      }),
    }),
    t.null,
  ]),
});
```

File: apps/nextjs-website/src/lib/fetch/types/CtaButton.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    variant: CTAButtonVariant,
+    color: CTAButtonColor,
+  }),
+  t.partial({
+    icon: t.union([t.string, t.null]),
+    size: CTAButtonSize,
+  }),
+]);
+
+export type ctaButtons = t.TypeOf<typeof CTAButtonSchema>;
```
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/types/CtaButton.ts:38]: Never used

```suggestion
```

File: apps/nextjs-website/src/lib/fetch/types/CtaButton.ts
Comment (datalek) [inline apps/nextjs-website/src/lib/fetch/types/CtaButton.ts:1]: Keep the same naming convention (rename the file from `CtaButton` to `CTAButton`)

**Review by datalek — CHANGES_REQUESTED** [review]

File: apps/nextjs-website/src/lib/__tests__/header.test.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
+import { describe, it, expect } from 'vitest';
+import { makeHeaderProps } from '../header'; // Adjust the import path based on your file structure
```
[the remark this answers is not in the document]
Comment (kin0992) [reply apps/nextjs-website/src/lib/__tests__/header.test.ts:2]: 🤔

File: apps/nextjs-website/src/lib/header.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  theme: 'light' | 'dark'
+): MenuDropdownProp => {
+  const { title, path } = item;
+  const itemsArray = navigation
+    .filter(
+      (child) =>
+        child.parent && child.parent.id === item.id && child.menuAttached
+    )
+    .map((child) => ({
+      href: '/' + item.path + '/' + child.path,
```
Comment (kin0992) [inline apps/nextjs-website/src/lib/header.ts:18]:
```suggestion
      href: `/${item.path}/${child.path}`,
```

File: apps/nextjs-website/src/lib/header.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+      label: child.title,
+    }));
+
+  return {
+    theme,
+    label: title,
+    href: '/' + path,
```
Comment (kin0992) [inline apps/nextjs-website/src/lib/header.ts:25]:
```suggestion
    href: `/${path}`,
```

File: apps/nextjs-website/src/lib/__tests__/header.test.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+const header: Header = {
+  data: {
+    attributes: {
+      theme: 'light',
+      avatar: {
+        data: {
+          attributes: {
+            url: 'path/to/avatar.jpg',
+            alternativeText: 'Alt text', // Add this line if alternativeText is part of your data structure
```
Comment (kin0992) [inline apps/nextjs-website/src/lib/__tests__/header.test.ts:50]:
```suggestion
            alternativeText: 'Alt text',
```
is it a part of the structure or not? 🤔
do we want it to be tested or not?

File: apps/nextjs-website/src/lib/__tests__/header.test.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+};
+
+const childNavItem = {
+  order: 1,
+  id: 2,
+  title: 'Child',
+  path: 'child',
+  parent: parentNavItem,
+  menuAttached: true,
+};
```
Comment (kin0992) [inline apps/nextjs-website/src/lib/__tests__/header.test.ts:22]: move these common structures in a file called `data.ts` under the `__tests__` folder. These structures are used also in other tests

File: apps/nextjs-website/src/lib/__tests__/header.test.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+          label: 'Parent',
+          theme: 'light',
+          items: [
+            {
+              href: '/parent/child',
+              label: 'Child',
+            },
+          ],
+        },
+      ], // You may need to adjust this based on your specific implementation
```
Comment (kin0992) [inline apps/nextjs-website/src/lib/__tests__/header.test.ts:104]:
```suggestion
      ],
```

File: apps/nextjs-website/src/lib/__tests__/header.test.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
```
Comment (kin0992) [inline apps/nextjs-website/src/lib/__tests__/header.test.ts:135]:
```suggestion
      ],
```

**Review by kin0992 — CHANGES_REQUESTED** [review]

**Review by kin0992 — APPROVED** [review]

**Review by datalek — APPROVED** [review]
